### PR TITLE
#214 authorization improvements

### PR DESCRIPTION
**Pull request by mperk**
Resolve first item in #214
MethodInvocationAuthorizationService should check Roles

### CODE REVIEW

File: framework/src/Volo.Abp.Authorization/Volo/Abp/Authorization/MethodInvocationAuthorizationService.cs [hunk 1]
```diff
@@ -64,20 +64,26 @@ protected virtual IEnumerable<IAuthorizeData> GetAuthorizationDataAttributes(Met
...
+                await _authorizationService.CheckAsync(authorizationAttribute.Policy).ConfigureAwait(false);
+            }
+            else if (authorizationAttribute.Roles != null)
+            {
+                if(_currentUser.IsInRole(authorizationAttribute.Roles) == false)
                 {
-                    throw new AbpAuthorizationException("Authorization failed! User has not logged in.");
+                    throw new AbpAuthorizationException("Authorization failed! Given roles has not granted: " + authorizationAttribute.Roles);
                 }
             }
```
[the remark this answers is not in the document]
Comment (mperk) [reply framework/src/Volo.Abp.Authorization/Volo/Abp/Authorization/MethodInvocationAuthorizationService.cs]: ok thank you for quick reply. I will research and development

File: framework/test/Volo.Abp.Authorization.Tests/Volo/Abp/Authorization/Authorization_Tests.cs [hunk 2]
```diff
@@ -52,5 +52,30 @@ public void Should_Permission_Definition_GetGroup()
...
+                await _myAuthorizedService1.ProtectedByRole().ConfigureAwait(false);
+            }).ConfigureAwait(false);
+        }
+
+        [Fact]
+        public async Task Should_Allow_To_Call_Method_If_Has_No_Role_ProtectedByRole_Async()
+        {
+            int result = await _myAuthorizedService1.ProtectedByRole().ConfigureAwait(false);
+            result.ShouldBe(42);
+        }
```
Comment (maliming) [inline framework/test/Volo.Abp.Authorization.Tests/Volo/Abp/Authorization/Authorization_Tests.cs]:
```
await _myAuthorizedService1.ProtectedByRole().ConfigureAwait(false);
```
I guess one of the unit test methods will fail.

Comment (mperk) [reply framework/test/Volo.Abp.Authorization.Tests/Volo/Abp/Authorization/Authorization_Tests.cs]: Yes, it is fail. I am still developing for that. Had very little.

Comment (mperk) [reply framework/test/Volo.Abp.Authorization.Tests/Volo/Abp/Authorization/Authorization_Tests.cs]: Hi @maliming ,
I finished finally. You can check.

**Comment by maliming** [comment]
hi @mperk 
I made some changes. see https://github.com/abpframework/abp/pull/2659/commits/583cdbac1cba2c7ad756200d469bb4470f97294c

Regarding `AuthenticationSchemes`, if we use `Authorize `attribute in the web layer controller, the authentication function of Asp Net Core will automatically handle `Policy`, `Roles`, `AuthenticationSchemes`.

`MethodInvocationAuthorizationService `is more for application services.

Will `AuthenticationScheme `be used in application services? I'm not sure.

If we want to handle `AuthenticationSchemes`, we need to use `HttpContext`.

**Comment by mperk** [comment]
> hi @mperk
> I made some changes. see [583cdba](https://github.com/abpframework/abp/commit/583cdbac1cba2c7ad756200d469bb4470f97294c)
> 
> Regarding `AuthenticationSchemes`, if we use `Authorize `attribute in the web layer controller, the authentication function of Asp Net Core will automatically handle `Policy`, `Roles`, `AuthenticationSchemes`.
> 
> `MethodInvocationAuthorizationService `is more for application services.
> 
> Will `AuthenticationScheme `be used in application services? I'm not sure.
> 
> If we want to handle `AuthenticationSchemes`, we need to use `HttpContext`.

Very good. Thanks.

**Comment by hikalkan** [comment]
As I see, some unit tests are failing.

image


````
Message: 
    System.ArgumentNullException : Value cannot be null. (Parameter 'policy')
  Stack Trace: 
    AuthorizationServiceExtensions.AuthorizeAsync(IAuthorizationService service, ClaimsPrincipal user, Object resource, AuthorizationPolicy policy)
    AbpAuthorizationServiceExtensions.AuthorizeAsync(IAuthorizationService authorizationService, Object resource, AuthorizationPolicy policy) line 30
    AbpAuthorizationServiceExtensions.AuthorizeAsync(IAuthorizationService authorizationService, AuthorizationPolicy policy) line 39
    AbpAuthorizationServiceExtensions.IsGrantedAsync(IAuthorizationService authorizationService, AuthorizationPolicy policy) line 81
    AbpAuthorizationServiceExtensions.CheckAsync(IAuthorizationService authorizationService, AuthorizationPolicy policy) line 120
    MethodInvocationAuthorizationService.CheckAsync(MethodInvocationAuthorizationContext context) line 32
    AuthorizationInterceptor.AuthorizeAsync(IAbpMethodInvocation invocation) line 24
    AuthorizationInterceptor.InterceptAsync(IAbpMethodInvocation invocation) line 18
    CastleAsyncAbpInterceptorAdapter`1.InterceptAsync(IInvocation invocation, IInvocationProceedInfo proceedInfo, Func`3 proceed) line 20
    ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
    ResourceInvoker.<InvokeNextExceptionFilterAsync>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
    ResourceInvoker.Rethrow(ExceptionContextSealed context)
    ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
    ResourceInvoker.InvokeNextResourceFilter()
    --- End of stack trace from previous location where exception was thrown ---
    ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
    ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
    ResourceInvoker.InvokeFilterPipelineAsync()
    --- End of stack trace from previous location where exception was thrown ---
    ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
    EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
    AbpUnitOfWorkMiddleware.InvokeAsync(HttpContext context, RequestDelegate next) line 23
    <<UseMiddlewareInterface>b__1>d.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
    AbpExceptionHandlingMiddleware.InvokeAsync(HttpContext context, RequestDelegate next) line 32
    AbpExceptionHandlingMiddleware.InvokeAsync(HttpContext context, RequestDelegate next) line 52
    <<UseMiddlewareInterface>b__1>d.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
    AbpAuditingMiddleware.InvokeAsync(HttpContext context, RequestDelegate next) line 37
    AbpAuditingMiddleware.InvokeAsync(HttpContext context, RequestDelegate next) line 54
    <<UseMiddlewareInterface>b__1>d.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
    AuthorizationMiddleware.Invoke(HttpContext context)
    FakeAuthenticationMiddleware.InvokeAsync(HttpContext context, RequestDelegate next) line 29
    <<UseMiddlewareInterface>b__1>d.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
    RequestLocalizationMiddleware.Invoke(HttpContext context)
    AbpRequestLocalizationMiddleware.InvokeAsync(HttpContext context, RequestDelegate next) line 34
    <<UseMiddlewareInterface>b__1>d.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
    AbpCorrelationIdMiddleware.InvokeAsync(HttpContext context, RequestDelegate next) line 27
    <<UseMiddlewareInterface>b__1>d.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
    <<SendAsync>g__RunRequestAsync|0>d.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
    ClientHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
    AbpAspNetCoreTestBase`1.GetResponseAsync(String url, HttpStatusCode expectedStatusCode) line 46
    AbpAspNetCoreTestBase`1.GetResponseAsStringAsync(String url, HttpStatusCode expectedStatusCode) line 35
    AuthTestController_Tests.Should_Authorize_For_Defined_And_Allowed_Permission() line 84
    --- End of stack trace from previous location where exception was thrown ---

````

**Comment by maliming** [comment]
@hikalkan  I fixed the unit tests.

**Comment by hikalkan** [comment]
Thanks to both of you :)

**Pull request merged**